index.js

```diff
@@ -3044,81 +3044,150 @@
     impl: at
   };
 
-  //# head :: Array a -> Maybe a
+  //# head :: Foldable f => f a -> Maybe a
   //.
-  //. Returns Just the first element of the given array if the array contains
-  //. at least one element; Nothing otherwise.
+  //. Returns Just the first element of the given structure if it
+  //. contains at least one element; Nothing otherwise.
   //.
   //. ```javascript
   //. > S.head ([1, 2, 3])
   //. Just (1)
   //.
   //. > S.head ([])
   //. Nothing
+  //.
+  //. > S.head (Cons ('foo') (Cons ('bar') (Nil)))
+  //. Just ('foo')
+  //.
+  //. > S.head (Nil)
+  //. Nothing
   //. ```
+
+  function head(foldable) {
+    return Z.reduce (
+      function(maybe, x) {
+        return maybe.isJust ? maybe : Just (x);
+      },
+      Nothing,
+      foldable
+    );
+  }
   _.head = {
-    consts: {},
-    types: [$.Array (a), $Maybe (a)],
-    impl: array (Nothing) (B (K) (Just))
+    consts: {f: [Z.Foldable]},
+    types: [f (a), $Maybe (a)],
+    impl: head
+
   };
 
-  //# last :: Array a -> Maybe a
+  //# last :: Foldable f => f a -> Maybe a
   //.
-  //. Returns Just the last element of the given array if the array contains
-  //. at least one element; Nothing otherwise.
+  //. Returns Just the last element of the given structure if it
+  //. contains at least one element; Nothing otherwise.
   //.
   //. ```javascript
   //. > S.last ([1, 2, 3])
   //. Just (3)
   //.
   //. > S.last ([])
   //. Nothing
+  //.
+  //. > S.last (Cons ('foo') (Cons ('bar') (Cons ('baz') (Nil))))
+  //. Just ('baz')
+  //.
+  //. > S.last (Nil)
+  //. Nothing
   //. ```
-  function last(xs) {
-    return xs.length > 0 ? Just (xs[xs.length - 1]) : Nothing;
+
+  function last(foldable) {
+    return Z.reduce (
+      function(maybe, x) {
+        return Just (x);
+      },
+      Nothing,
+      foldable
+     );
   }
+
   _.last = {
-    consts: {},
-    types: [$.Array (a), $Maybe (a)],
+    consts: {f: [Z.Foldable]},
+    types: [f (a), $Maybe (a)],
     impl: last
   };
 
-  //# tail :: Array a -> Maybe (Array a)
+  //# tail :: (Applicative f, Foldable f, Monoid (f a)) => f a -> Maybe (f a)
   //.
-  //. Returns Just all but the first of the given array's elements if the
-  //. array contains at least one element; Nothing otherwise.
+  //. Returns a Just of all but the first of the given structure if
+  //. it contains at least one element; Nothing otherwise.
   //.
   //. ```javascript
   //. > S.tail ([1, 2, 3])
   //. Just ([2, 3])
   //.
   //. > S.tail ([])
   //. Nothing
+  //.
+  //. > S.tail (Cons ('foo') (Cons ('bar') (Cons ('baz') (Nil))))
+  //. Just (Cons ('bar') (Cons ('baz') (Nil)))
+  //
+  //. > S.tail (Nil)
+  //. Nothing
   //. ```
+
+  function tail(foldable) {
+    return Z.reduce (
+      function(maybe, x) {
+        return maybe.isNothing ?
+               Just (Z.empty (foldable.constructor)) :
+               Z.map (append (x), maybe);
```

Comment:
Something doesn't feel right about using `Z.map` in conjunction with the `isNothing` check. If we're to keep the `isNothing` check we could safely use `maybe.value` in the Just case:

```javascript
return maybe.isNothing ?
       Just (Pair (Z.empty (foldable.constructor)) (x)) :
       Just (Pair (Z.append (maybe.value.snd, maybe.value.fst)) (x));
```

This could be improved by removing the common code from the branches:

```javascript
return Just (maybe.isNothing ?
             Pair (Z.empty (foldable.constructor)) (x) :
             Pair (Z.append (maybe.value.snd, maybe.value.fst)) (x));
```

```javascript
return Just (Pair (maybe.isNothing ?
                   Z.empty (foldable.constructor) :
                   Z.append (maybe.value.snd, maybe.value.fst))
                  (x));
```

We could even remove the explicit conditional logic by using `Z.reduce`:

```javascript
return Just (Pair (Z.reduce (function(_, pair) { return Z.append (pair.snd, pair.fst); },
                             Z.empty (foldable.constructor),
                             maybe))
                  (x));
```

The problem with this approach is that `Z.empty (foldable.constructor)` would be evaluated once for each element of the structure. Can you think of a solution to this problem?

Reply:
The comment above applies to `init` rather than to `tail`. I'm sorry for causing confusion.

Reply:
I stored the empty in a var.